version 1.0

# map using bowtie2
task Align {
    input {
      Boolean paired_end_run
      Boolean directional
      File fastq_input_read_a
      File? fastq_input_read_b
      File reference_fasta
      File reference_fasta_index
      File fwd_converted_reference_fasta
      Array[File] fwd_bowtie2_index_files
      File rev_converted_reference_fasta
      Array[File] rev_bowtie2_index_files
      String output_base_name
     
    }

      # output name for mapped reads
      # output filenames are generated by Bismark
      # if/else used for paired end vs sinlge end alignment
      String bismark_mapped_bam_output_name = basename(fastq_input_read_a, ".fastq.gz") +
                                        (if paired_end_run then "_bismark_bt2_pe.bam" else "_bismark_bt2.bam")
      String bismark_mapping_report_output_name = basename(fastq_input_read_a, ".fastq.gz") +
                                            (if paired_end_run then "_bismark_bt2_PE_report.txt" else "_bismark_bt2_SE_report.txt")
      String mapped_bam_output_name = output_base_name + ".aligned.bam"
      String mapping_report_output_name = output_base_name + ".alignment_report.txt"

      # input file sizes
      Float reference_size = size(reference_fasta, "GB") +
                                size(fwd_converted_reference_fasta, "GB") +
                                size(rev_converted_reference_fasta, "GB")
      Float bam_size = size(fastq_input_read_a, "GB") +
                               (if paired_end_run then size(fastq_input_read_b, "GB") else 0)

  # align with bowtie2
  command <<<
    set -euo pipefail


    # create dir struture required by bismark
    # move input files into corresponding sub dir
    declare -a REF_DIR=$(mktemp -d genome_refXXXXXX)
    mkdir $REF_DIR/Bisulfite_Genome
    mkdir $REF_DIR/Bisulfite_Genome/CT_conversion
    mkdir $REF_DIR/Bisulfite_Genome/GA_conversion

    mv ~{reference_fasta} $REF_DIR
    mv ~{reference_fasta_index} $REF_DIR
    mv ~{fwd_converted_reference_fasta} $REF_DIR/Bisulfite_Genome/CT_conversion
    mv ~{sep=' ' fwd_bowtie2_index_files} $REF_DIR/Bisulfite_Genome/CT_conversion
    mv ~{rev_converted_reference_fasta} $REF_DIR/Bisulfite_Genome/GA_conversion
    mv ~{sep=' ' rev_bowtie2_index_files} $REF_DIR/Bisulfite_Genome/GA_conversion

    # use bismark with bowtie2 to align the read
    # icpc to not attach read ID descriptions to read ID
      bismark $REF_DIR/ \
      --bowtie2 \
      --icpc \
      ~{true="-1" false="" paired_end_run} ~{fastq_input_read_a} \
      ~{true="-2" false="" paired_end_run} ~{fastq_input_read_b} \
      ~{true="-X 2000" false="" paired_end_run} \
      ~{true="--pbat" false="" directional}

    # rename defualt bismark output name to match basename
    mv ~{bismark_mapped_bam_output_name} ~{mapped_bam_output_name}
    mv ~{bismark_mapping_report_output_name} ~{mapping_report_output_name}
  >>>

  runtime {
    docker: "quay.io/biocontainers/bismark:0.24.2--hdfd78af_0"
    cpu: 4
    
  }

  output {
    File mapped_bam_output = mapped_bam_output_name
    File mapping_report_output = mapping_report_output_name
    
  }
}
